src/features/contracts/index.ts

```diff
@@ -0,0 +1,196 @@
+import * as _ from 'lodash';
+import * as memoizee from 'memoizee';
+import * as Bluebird from 'bluebird';
+import { fetchContractsLocally, getContracts } from './contracts-directory';
+import { sbvrUtils, permissions } from '@balena/pinejs';
+import {
+	CONTRACTS_PRIVATE_REPO_NAME,
+	CONTRACTS_PRIVATE_REPO_OWNER,
+	CONTRACTS_PRIVATE_REPO_TOKEN,
+	CONTRACTS_PUBLIC_REPO_NAME,
+	CONTRACTS_PUBLIC_REPO_OWNER,
+} from '../../lib/config';
+
+const SYNC_INTERVAL = 5 * 60 * 1000;
+
+export interface RepositoryInfo {
+	owner: string;
+	name: string;
+	token?: string;
+}
+
+export interface Contract {
+	slug: string;
+	version: number;
+	type: string;
+	aliases?: string[];
+	name: string;
+	data: any;
+}
+
+export const getContractRepos = (): RepositoryInfo[] => {
+	const repos: RepositoryInfo[] = [
+		{
+			owner: CONTRACTS_PUBLIC_REPO_OWNER,
+			name: CONTRACTS_PUBLIC_REPO_NAME,
+		},
+	];
+
+	if (
+		CONTRACTS_PRIVATE_REPO_OWNER &&
+		CONTRACTS_PRIVATE_REPO_NAME &&
+		CONTRACTS_PRIVATE_REPO_TOKEN
+	) {
+		repos.push({
+			owner: CONTRACTS_PRIVATE_REPO_OWNER,
+			name: CONTRACTS_PRIVATE_REPO_NAME,
+			token: CONTRACTS_PRIVATE_REPO_TOKEN,
+		});
+	}
+
+	return repos;
+};
+
+type FieldsMap = {
+	[dbField: string]: {
+		contractField: string;
+		default?: any;
+	};
+};
+// This map will hold information on which contract fields imported from the contract type will be synced to which db model and fields.
+type SyncSettings = {
+	[contractType: string]: {
+		resource: string;
+		uniqueKey: string;
+		map: FieldsMap;
+	};
+};
+
+let globalSyncSettings: SyncSettings | undefined;
+
+export function setSyncSettings(syncSettings: SyncSettings) {
+	globalSyncSettings = syncSettings;
+}
+
+const mapModel = (contractEntry: Contract, map: FieldsMap) => {
+	const mappedModel: { [k: string]: any } = {};
+	for (const key of Object.keys(map) as Array<keyof typeof map>) {
+		mappedModel[key] =
+			_.get(contractEntry, map[key]?.contractField) ?? map[key]!.default;
+	}
+
+	return mappedModel;
+};
+
+const upsertEntries = async (
+	newData: any[],
+	existingData: Set<string | number | boolean>,
+	resource: string,
+	uniqueField: string,
+	rootApi: sbvrUtils.PinejsClient,
+) => {
+	await Bluebird.map(
+		newData,
+		async (entry: any) => {
+			try {
+				if (existingData.has(entry[uniqueField])) {
+					return await rootApi.patch({
+						resource,
+						body: entry,
+						options: {
+							$filter: { slug: entry[uniqueField] },
```

Comment:
Here you use `slug` for the db-side unique field but elsewhere you use `uniqueField` as the db-side field. I'm assuming that the unique field in the contract happens to be "slug" - could you do a test with the unique field being "foobar" and setting that in the mapping as I expect it'll catch some issues where you assume they have to match

Reply:
Yeah I missed that one, nice catch.